Allow action to return multiple groups of files, based on their status

apostroll opened the pull request:
This is a breaking change!

The action will now return 3 groups of files:

* `all_changed_files`: A list of all files, new, modified or removed
* `new_or_modified_files`: A list of new or modified files. Removed files are excluded here.
* `removed_files`: A list of removed files only!

To simplify the implementation, we only support single include and exclude regular expressions. Did you know you can combine your regular expressions?!!!

Additionally, the count of modified files was removed, as it was deemed unnecessary. You can count on your own after all!

apostroll committed:
Allow action to return multiple groups of files, based on their status

This is a breaking change!

The action will now return 3 groups of files:

* `all_changed_files`: A list of all files, new, modified or removed
* `new_or_modified_files`: A list of new or modified files. Removed files are excluded here.
* `removed_files`: A list of removed files only!

To simplify the implementation, we only support single include and exclude regular expressions. Did you know you can combine your regular expressions?!!!

Additionally, the count of modified files was removed, as it was deemed unnecessary. You can count on your own after all!

apostroll commented:
This was used on the `aws-infra` workflow, on [this PR](https://github.com/skroutz-internal/aws-infra/pull/1743) in order to fix an issue where, the `aws-infra` workflow failed to run because the PR appeared to be a NOOP, causing GitHub to block merges.

apostroll commented on action.yml:
```diff
@@ -29,52 +32,53 @@ runs:
-            echo "Compare API returned an error. Falling back to git diff."
-            FILES=$(git diff --name-only "${{ github.event.before }}" "${{ github.sha }}" || true)
-          fi
+          response=$(gh api repos/${{ github.repository }}/compare/${{ github.event.before }}...${{ github.sha }}/files)
```
I consider this an acceptable outcome. If the GH api fails, it probably means several other things are going wrong at the same time.

Will revisit this if necessary.

phoinixgrr commented:
Thanks @apostroll for contributing :rock: 

https://github.com/search?q=org%3Askroutz-internal%20skroutz%2Faction-changed-files&type=code

This is used in multiple places. No objections merging this, as long as the breking changes introduced do not affect the dependant workflows.

phoinixgrr commented:
Let's also update the Readme.MD file accordingly

apostroll commented:
> Thanks @apostroll for contributing 🪨
> 
> https://github.com/search?q=org%3Askroutz-internal%20skroutz%2Faction-changed-files&type=code
> 
> This is used in multiple places. No objections merging this, as long as the breking changes introduced do not affect the dependant workflows.

The idea is to track it via git branch, in order to avoid breaking changes to other workflows.

apostroll commented:
I've kept the `v2` branch to be used as a pointer, and also registered the `v2.0.0.` tag.

No real need to merge this on main if it's to cause headaches.

phoinixgrr commented:
I’d recommend merging the changes into the main branch to avoid creating two diverging versions of the same software.

Additionally, the current release of the action has already been modified in a non-standard way. According to our convention, releases **should be made from the main branch**, not from feature branches. This is to have a stable and expected release of software versioning. 

If we intend to maintain this divergence, I’d suggest dropping the v2.0 release and instead referencing the specific commit hash from your branch directly.

phoinixgrr reviewed (changes requested):
I’d recommend merging the changes into the main branch, after making shure, that the new release, is either compatible for the repositories using it, or changes are made to the repositorires using it, to avoid creating two diverging versions of the same software.

Additionally, the current release of the action has already been modified in a non-standard way. According to our convention, releases **should be made from the main branch**, not from feature branches. This is to have a stable and expected release of software versioning. 

If we intend to maintain this divergence, I’d suggest dropping the v2.0 release and instead referencing the specific commit hash from your branch directly.

apostroll committed:
Bug fixes

Fixes invalid gh api URL used to compare 2 commits at the `/files` suffix will always return 404
(see: [GitHub CLI compare two commits](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#compare-two-commits))

Fixes invalid quotation on several expressions used that we never noticed, writing shell scripts inside YAML (that should be a no go in the future).

apostroll commented:
> I’d recommend merging the changes into the main branch, after making shure, that the new release, is either compatible for the repositories using it, or changes are made to the repositorires using it, to avoid creating two diverging versions of the same software.
> 
> Additionally, the current release of the action has already been modified in a non-standard way. According to our convention, releases **should be made from the main branch**, not from feature branches. This is to have a stable and expected release of software versioning.
> 
> If we intend to maintain this divergence, I’d suggest dropping the v2.0 release and instead referencing the specific commit hash from your branch directly.

Removed the v2.0.0 & v2.0.1 tags & releases, so that we can recreate them on the main branch
